File: DoMemory/DoMemory/Services/GameStats/ProfileStatsService.swift
//
//  ProfileStatsService.swift
//  DoMemory
//

import Foundation

struct ProfileStats {
    let totalPlayed: Int
    let totalWon: Int
    let perfectGames: Int
    let multiplayerWins: Int
    let longestStreak: Int

    var winRate: Double {
        totalPlayed > 0 ? Double(totalWon) / Double(totalPlayed) : 0
    }
}

struct Achievement: Identifiable {
    let id: String
    let title: String
    let detail: String
    let symbol: String
    let isUnlocked: Bool
    /// 0...1 progress toward unlocking.
    let progress: Double
}

/// Lifetime aggregate stats and derived achievements, backed by `UserDefaults`.
/// Mirrors `GameStatsService` (per-game) but tracks global totals.
final class ProfileStatsService {
    static let shared = ProfileStatsService()

    private let defaults = UserDefaults.standard
    private init() {}

    private enum Key {
        static let played = "profile.totalPlayed"
        static let won = "profile.totalWon"
        static let perfect = "profile.perfectGames"
        static let multiplayerWins = "profile.multiplayerWins"
        static func bestRemaining(_ difficulty: Difficulty) -> String { "profile.bestRemaining.\(difficulty.rawValue)" }
    }

    // MARK: - Recording

    func recordGameFinished(didWin: Bool, isPerfect: Bool, difficulty: Difficulty, timeRemaining: Int) {
        defaults.set(defaults.integer(forKey: Key.played) + 1, forKey: Key.played)
        guard didWin else { return }
        defaults.set(defaults.integer(forKey: Key.won) + 1, forKey: Key.won)
        if isPerfect {
            defaults.set(defaults.integer(forKey: Key.perfect) + 1, forKey: Key.perfect)
        }
        let bestKey = Key.bestRemaining(difficulty)
        if timeRemaining > defaults.integer(forKey: bestKey) {
            defaults.set(timeRemaining, forKey: bestKey)
        }
    }

    func recordMultiplayerWin() {
        defaults.set(defaults.integer(forKey: Key.multiplayerWins) + 1, forKey: Key.multiplayerWins)
    }

    /// Highest remaining time (= fastest win) recorded for a difficulty, or nil if none.
    func bestRemaining(for difficulty: Difficulty) -> Int? {
        let value = defaults.integer(forKey: Key.bestRemaining(difficulty))
        return value > 0 ? value : nil
    }

    // MARK: - Reading

    var stats: ProfileStats {
        ProfileStats(
            totalPlayed: defaults.integer(forKey: Key.played),
            totalWon: defaults.integer(forKey: Key.won),
            perfectGames: defaults.integer(forKey: Key.perfect),
            multiplayerWins: defaults.integer(forKey: Key.multiplayerWins),
            longestStreak: DailyChallengeService.shared.longestStreak
        )
    }

    func achievements() -> [Achievement] {
        let stats = self.stats

        func winTier(_ threshold: Int) -> Achievement {
            Achievement(
                id: "wins_\(threshold)",
                title: String(format: NSLocalizedString("ach_wins_title_format", comment: "Wins achievement title, %d = count"), threshold),
                detail: String(format: NSLocalizedString("ach_wins_detail_format", comment: "Wins achievement detail, %d = count"), threshold),
                symbol: "rosette",
                isUnlocked: stats.totalWon >= threshold,
                progress: min(1, Double(stats.totalWon) / Double(threshold))
            )
        }

        func streakTier(_ threshold: Int) -> Achievement {
            Achievement(
                id: "streak_\(threshold)",
                title: String(format: NSLocalizedString("ach_streak_title_format", comment: "Streak achievement title, %d = days"), threshold),
                detail: String(format: NSLocalizedString("ach_streak_detail_format", comment: "Streak achievement detail, %d = days"), threshold),
                symbol: "flame.fill",
                isUnlocked: stats.longestStreak >= threshold,
                progress: min(1, Double(stats.longestStreak) / Double(threshold))
            )
        }

        let perfect = Achievement(
            id: "perfect",
            title: NSLocalizedString("ach_perfect_title", comment: "Perfect game achievement title"),
            detail: NSLocalizedString("ach_perfect_detail", comment: "Perfect game achievement detail"),
            symbol: "sparkles",
            isUnlocked: stats.perfectGames > 0,
            progress: stats.perfectGames > 0 ? 1 : 0
        )

        let multiplayer = Achievement(
            id: "multiplayer",
            title: NSLocalizedString("ach_multiplayer_title", comment: "Multiplayer achievement title"),
            detail: NSLocalizedString("ach_multiplayer_detail", comment: "Multiplayer achievement detail"),
            symbol: "person.2.fill",
            isUnlocked: stats.multiplayerWins > 0,
            progress: stats.multiplayerWins > 0 ? 1 : 0
        )

        return [winTier(10), winTier(50), winTier(100), streakTier(7), streakTier(30), perfect, multiplayer]
    }
}

File: android/app/src/main/java/com/ezequielbrrt/domemory/services/stats/ProfileStatsService.kt
package com.ezequielbrrt.domemory.services.stats

import com.ezequielbrrt.domemory.R
import com.ezequielbrrt.domemory.data.prefs.UserPreferences
import com.ezequielbrrt.domemory.services.daily.DailyChallengeService
import kotlinx.coroutines.flow.first

/** Lifetime aggregate stats, mirroring iOS's `ProfileStats`. */
data class ProfileStats(
    val totalPlayed: Int = 0,
    val totalWon: Int = 0,
    val perfectGames: Int = 0,
    val multiplayerWins: Int = 0,
    val longestStreak: Int = 0,
) {
    val winRate: Float get() = if (totalPlayed > 0) totalWon.toFloat() / totalPlayed else 0f
}

/**
 * One derived badge. [titleRes]/[detailRes] are string resources rather than raw text —
 * both the win and streak tiers share one `%d`-format resource
 * ([android.text.format]-style), passed as [formatArg]; the perfect-game and multiplayer
 * badges take neither, so [formatArg] stays null for them. [emoji] stands in for iOS's SF
 * Symbol (`symbol`) — this app has no icon library and none should be added for this (see
 * `ANDROID_PLAN.md`'s Phase 8 note); the locked visual state is rendered as 🔒 by the
 * composable, not stored here, mirroring iOS's own `achievement.isUnlocked ? achievement
 * .symbol : "lock.fill"` swap at the view layer.
 */
data class Achievement(
    val id: String,
    val titleRes: Int,
    val detailRes: Int,
    val formatArg: Int? = null,
    val emoji: String,
    val isUnlocked: Boolean,
    /** 0f..1f progress toward unlocking, capped at 1f. */
    val progress: Float,
)

/**
 * Lifetime aggregates and derived achievements, backed by [UserPreferences] (spec 13.2).
 * Mirrors iOS's `ProfileStatsService`, minus the singleton shape — this is constructed once
 * in `AppContainer`, like every other service here (decision D4).
 *
 * [UserPreferences] has no synchronous read (unlike iOS's `UserDefaults`), so [stats] is a
 * suspend snapshot rather than a live value — `AchievementsViewModel` takes one snapshot per
 * screen visit, the same "explicit pull" convention `DailyChallengeService`'s own readers
 * already use, rather than a continuously-collected combine of five flows.
 */
class ProfileStatsService(
    private val prefs: UserPreferences,
    private val dailyChallenge: DailyChallengeService,
) {
    suspend fun stats(): ProfileStats = ProfileStats(
        totalPlayed = prefs.totalPlayed.first(),
        totalWon = prefs.totalWon.first(),
        perfectGames = prefs.perfectGames.first(),
        multiplayerWins = prefs.multiplayerWins.first(),
        longestStreak = dailyChallenge.longestStreak(),
    )

    suspend fun achievements(): List<Achievement> = achievements(stats())

    companion object {
        val WIN_TIERS = listOf(10, 50, 100)
        val STREAK_TIERS = listOf(7, 30)

        /**
         * The pure derivation — given a [ProfileStats] snapshot, no [UserPreferences] or
         * coroutine involved, so it is exactly as unit-testable as iOS's own
         * `achievements()` (which reads `self.stats`, a synchronous `UserDefaults`
         * snapshot, and does the rest with no I/O either). Matches iOS's ordering and
         * thresholds exactly: three win tiers, two streak tiers, one perfect-game badge,
         * one multiplayer badge.
         */
        fun achievements(stats: ProfileStats): List<Achievement> {
            val winTiers = WIN_TIERS.map { threshold ->
                Achievement(
                    id = "wins_$threshold",
                    titleRes = R.string.ach_wins_title_format,
                    detailRes = R.string.ach_wins_detail_format,
                    formatArg = threshold,
                    emoji = "🏅",
                    isUnlocked = stats.totalWon >= threshold,
                    progress = (stats.totalWon.toFloat() / threshold).coerceIn(0f, 1f),
                )
            }
            val streakTiers = STREAK_TIERS.map { threshold ->
                Achievement(
                    id = "streak_$threshold",
                    titleRes = R.string.ach_streak_title_format,
                    detailRes = R.string.ach_streak_detail_format,
                    formatArg = threshold,
                    emoji = "🔥",
                    isUnlocked = stats.longestStreak >= threshold,
                    progress = (stats.longestStreak.toFloat() / threshold).coerceIn(0f, 1f),
                )
            }
            val perfect = Achievement(
                id = "perfect",
                titleRes = R.string.ach_perfect_title,
                detailRes = R.string.ach_perfect_detail,
                emoji = "✨",
                isUnlocked = stats.perfectGames > 0,
                progress = if (stats.perfectGames > 0) 1f else 0f,
            )
            val multiplayer = Achievement(
                id = "multiplayer",
                titleRes = R.string.ach_multiplayer_title,
                detailRes = R.string.ach_multiplayer_detail,
                emoji = "🏆",
                isUnlocked = stats.multiplayerWins > 0,
                progress = if (stats.multiplayerWins > 0) 1f else 0f,
            )
            return winTiers + streakTiers + listOf(perfect, multiplayer)
        }
    }
}
